### ISSUE TITLE
Make perfkit code independent for providers code

### ISSUE THREAD

**Issue by kivio** [GoogleCloudPlatform/PerfKitBenchmarker#624]
What does mean for me:
1. Remove provider configuration from benchmark_spec
2. Remove dependencies on providers from benchmarks (https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/issues/621)

I don't known what more but i can propose whole solution for that soon - @voellm what do you think about that?
It's big change and doing it step by step can takes big amount of time? Then my propose is to do it in one part, as fast as fast is possible, i can propose something before monday.

**Comment by voellm** [comment]
What are you trying to achieve here?  Not really sure I understand from the description here.

**Comment by kivio** [comment]
In future look i want to have installed just that providers than i want to use.

In shortest, i have problem because when i was updating my sourcecode of PKB i catch error.
"c(loud)s(stack)api not found" - wait a moment i want to run openstack provider.

Ok this looks like problem with cloudstack provider, i was using novaclient in my driver but i was doing it more smartly and you catch this error just if you haven't novaclient when you try to run tests on openstack. But it is not problem with cloudstack driver, we can describe this problem that way. When you want to run benchmark, code of all providers is importing. I think it is bad way. 

I want to make core of perfkitbenchmarker more providers/benchmarks/packages/tracers and other things like that independent.

**Comment by voellm** [comment]
Agreed... @connormccoy noted this as well.  Our plan is to allow the code
to go together but delay load only when you use it.

@noahl -
We should have this addressed in the next release v0.24.0

**Comment by kivio** [comment]
@voellm - it's not finished but you can look here (https://github.com/kivio/PerfKitBenchmarker/tree/refactor_providers) in provider package for my conception of make this change

**Comment by noahl** [comment]
@ehankland Would you please assign this to yourself, since you're working on the fix?

**Comment by ehankland** [comment]
Sure - I'm not sure what I'm working on will completely fix this issue though - there's still the matter of moving object storage and sql service provider code into the provider directories.

**Comment by noahl** [comment]
Oh, I was thinking of this bug as just covering not loading provider APIs we don't need. That's what I think we need to merge for this release. We can track the rest of the work in #621 .

**Comment by noahl** [comment]
Eric just merged the part of this that I wanted to get done in 0.24.0. I'm going to remove this from the 0.24.0 milestone.

**Comment by ehankland** [comment]
Should we keep this open to help track the remaining work or close it now that providers are loaded only if they are needed?

**Comment by voellm** [comment]
@ehankland 
@noahl 

is there still remaining work?  If so can we please list it all here.

**Comment by kivio** [comment]
@noahl - #621 is good part of what i meen, but not all. I think more about dynamic dependencies injection. Importing provider just where is need and auto-discover options of them like Networks, Disks, VMs and others. All for remove configuration of providers dict from benchmark_spec. I think i can show you code after weekend.

**Comment by ehankland** [comment]
@kivio We recently made some changes that might be what you are looking for - the providers dict is gone from benchmark_spec.py and we do only load providers when they are used now. Can you take a look to see if there's more you wanted done?
